File: packages/contracts/src/types/menu.ts
// Menu catalog contracts — added at 0.2.2. Mirrors go/menu.go.
//
// These are CONFIG aggregates under §50.1: the cloud owns them, they sync
// cloud→edge versioned by config_version, and the edge replaces them wholesale
// rather than merging. They already carry their authority direction in
// AGGREGATE_AUTHORITY (menu_item since 0.1.0); this file exists so the POS
// frontend and the Tauri layer share one shape instead of each hand-rolling
// the SQLite column names.
//
// Field names and types match packages/contracts/sqlite/0001_init.sql and
// postgres/0001_init.sql exactly. Money is integer paise, never float.

import { z } from "zod";

export const MenuCategorySchema = z.object({
  id: z.string().uuid(),
  outlet_id: z.string().uuid(),
  name: z.string(),
  sort_order: z.number().int(),
  config_version: z.number().int(),
  schema_version: z.literal(1),
});
export type MenuCategory = z.infer<typeof MenuCategorySchema>;

export const MenuItemSchema = z.object({
  id: z.string().uuid(),
  outlet_id: z.string().uuid(),
  category_id: z.string().uuid(),
  name: z.string(),
  base_price_paise: z.number().int(), // integer paise (CLAUDE.md); never float
  is_available: z.boolean(), // item-snooze flag, §19
  // Added at 0.4.2 (ADR-016 addendum). Which TaxProfile prices this item.
  //
  // NULL is meaningful: "use the outlet's default profile". That keeps a
  // single-rate restaurant configuration-free rather than making every item
  // name the same profile.
  //
  // This is an INPUT to resolution, never a substitute for the snapshot.
  // Resolution happens at billing time and invoice_line stores what was
  // applied — its own tax_profile_id plus per-component rate_bps and paise.
  // Re-pointing an item at a different profile tomorrow must never alter what
  // a bill issued today says it charged (§31 reproducibility).
  tax_profile_id: z.string().uuid().nullable(),
  // Added at 0.4.5 (ADR-016 addendum). The HSN (goods) or SAC (services) code
  // a GST tax invoice must print for this item's lines.
  //
  // Lives here rather than on TaxProfile because HSN/SAC classifies WHAT IS
  // SOLD while a profile classifies HOW IT IS RATED: prepared food (SAC 9963)
  // and packaged bottled water (HSN 2201) routinely share one 5% profile, so
  // hanging the code off the profile would force them to share a code.
  //
  // NULLABLE, and not a licence to print a blank code. Nullable because this
  // was additive over catalogues that had none, and because a default of
  // '9963' for everything would stamp a plausible, wrong, legally-meaningful
  // code on packaged goods — a wrong HSN is worse than a missing one, because
  // it looks configured. The completeness rule lives at ISSUE time: an invoice
  // must not issue with a line whose hsn_sac is NULL, and the error must name
  // the item so the catalogue can be fixed.
  //
  // Like tax_profile_id, an INPUT to resolution and never a substitute for the
  // snapshot — invoice_line.hsn_sac stores what was applied, so correcting a
  // catalogue tomorrow never rewrites a bill issued today (§31).
  hsn_sac: z.string().nullable(),
  config_version: z.number().int(),
  schema_version: z.literal(1),
});
export type MenuItem = z.infer<typeof MenuItemSchema>;

export const MenuItemVariantSchema = z.object({
  id: z.string().uuid(),
  menu_item_id: z.string().uuid(),
  name: z.string(),
  price_delta_paise: z.number().int(),
  // 0.5.7 — landed at 0.5.0 in the store (sqlite/0014) and never on the wire.
  is_default: z.boolean(),
  config_version: z.number().int(),
  schema_version: z.literal(1),
});
export type MenuItemVariant = z.infer<typeof MenuItemVariantSchema>;

export const MenuItemModifierSchema = z.object({
  id: z.string().uuid(),
  menu_item_id: z.string().uuid(),
  group_name: z.string(),
  option_name: z.string(),
  price_delta_paise: z.number().int(),
  min_selection: z.number().int(),
  max_selection: z.number().int(),
  config_version: z.number().int(),
  schema_version: z.literal(1),
});
export type MenuItemModifier = z.infer<typeof MenuItemModifierSchema>;

File: packages/contracts/go/menu.go
// Menu catalog contracts — added at 0.2.2. Mirrors src/types/menu.ts.
//
// CONFIG aggregates under §50.1: cloud-owned, synced cloud→edge versioned by
// config_version, replaced wholesale at the edge rather than merged.
// Field names match sqlite/0001_init.sql and postgres/0001_init.sql exactly.
// Money is integer paise, never float.
package contracts

type MenuCategory struct {
	ID            string `json:"id"`
	OutletID      string `json:"outlet_id"`
	Name          string `json:"name"`
	SortOrder     int    `json:"sort_order"`
	ConfigVersion int    `json:"config_version"`
	SchemaVersion int    `json:"schema_version"`
}

type MenuItem struct {
	ID             string `json:"id"`
	OutletID       string `json:"outlet_id"`
	CategoryID     string `json:"category_id"`
	Name           string `json:"name"`
	BasePricePaise int64  `json:"base_price_paise"`
	IsAvailable    bool   `json:"is_available"`
	// TaxProfileID added at 0.4.2 (ADR-016 addendum). Which TaxProfile prices
	// this item; nil means "use the outlet's default profile".
	//
	// An INPUT to resolution, never a substitute for the snapshot: resolution
	// happens at billing time and InvoiceLine stores what was applied (its own
	// TaxProfileID plus per-component rate_bps and paise), so re-pointing an
	// item tomorrow never rewrites a bill issued today (§31).
	TaxProfileID *string `json:"tax_profile_id"`
	// HSNSAC added at 0.4.5 (ADR-016 addendum). The HSN (goods) or SAC
	// (services) code a GST tax invoice must print for this item's lines.
	//
	// On MenuItem rather than TaxProfile because HSN/SAC classifies what is
	// sold while a profile classifies how it is rated — prepared food
	// (SAC 9963) and packaged water (HSN 2201) routinely share one 5%
	// profile, so hanging the code off the profile would force them to
	// share a code.
	//
	// Nil-able, and not a licence to print a blank code: a default of
	// "9963" for everything would stamp a plausible, wrong, legally
	// meaningful code on packaged goods, and a wrong HSN is worse than a
	// missing one because it looks configured. The completeness rule lives
	// at invoice-issue time, which must reject a NULL line and name the
	// item.
	//
	// An INPUT to resolution: InvoiceLine.HSNSAC stores what was applied,
	// so a catalogue correction never rewrites an issued bill (§31).
	HSNSAC        *string `json:"hsn_sac"`
	ConfigVersion int     `json:"config_version"`
	SchemaVersion int     `json:"schema_version"`
}

type MenuItemVariant struct {
	ID              string `json:"id"`
	MenuItemID      string `json:"menu_item_id"`
	Name            string `json:"name"`
	PriceDeltaPaise int64  `json:"price_delta_paise"`
	// 0.5.7. The column landed at 0.5.0 (sqlite/0014, ADR-018 §2.1) and the
	// wire types never got it -- the additive-consumer-list rule's failure
	// mode, predating the rule. Without it a default variant cannot sync, and
	// order lines at a cloud-synced outlet cannot stamp one.
	IsDefault     bool `json:"is_default"`
	ConfigVersion int  `json:"config_version"`
	SchemaVersion int  `json:"schema_version"`
}

type MenuItemModifier struct {
	ID              string `json:"id"`
	MenuItemID      string `json:"menu_item_id"`
	GroupName       string `json:"group_name"`
	OptionName      string `json:"option_name"`
	PriceDeltaPaise int64  `json:"price_delta_paise"`
	MinSelection    int    `json:"min_selection"`
	MaxSelection    int    `json:"max_selection"`
	ConfigVersion   int    `json:"config_version"`
	SchemaVersion   int    `json:"schema_version"`
}

// MenuItemPatch is the mutable field set of PATCH /menu/items/{itemId},
// added at 0.7.0 (ADR-024).
//
// EVERY FIELD IS A POINTER so "absent" and "set to the zero value" stay
// distinguishable: a nil Name leaves the name alone, while a non-nil pointer
// to "" is a caller error the handler rejects. Without pointers a PATCH that
// omitted a price would silently set it to zero, which on a money field is
// the worst available failure.
//
// IsAvailable is absent DELIBERATELY: POST /menu/items/{itemId}/availability
// owns it and is an EDGE->CLOUD replay route, so accepting it here would make
// the cloud a second writer of a field the outlet authors (§50.1).
type MenuItemPatch struct {
	Name           *string `json:"name,omitempty"`
	BasePricePaise *int64  `json:"base_price_paise,omitempty"`
	CategoryID     *string `json:"category_id,omitempty"`
	// Double pointer is deliberate: tax_profile_id is itself nullable, so the
	// three states are absent (leave alone), null (clear to the outlet
	// default), and set. A single pointer cannot express all three.
	TaxProfileID **string `json:"tax_profile_id,omitempty"`
	// May be changed, never cleared: an invoice cannot issue with a NULL or
	// blank HSN/SAC on any line (0.4.5).
	HSNSAC *string `json:"hsn_sac,omitempty"`
}
